File: client/src/services/audit-report.service.ts
import http from './http-common';
import PaginatedList from '../model/paginated-list.model';
import { AuditIssue, AuditIssueSummary, AuditReportConfig, AuditReportTemplate } from '../model/audit-report.model';

class AuditReportService {
  getReportTemplates(page: number = 0, pageSize: number = 1000): Promise<PaginatedList<AuditReportTemplate>> {
    return http.get<PaginatedList<AuditReportTemplate>>('/rails/audit/templates', { params: { "page": page, "page-size": pageSize }})
      .then(response => response.data);
  }

  async fetchAllTemplates(): Promise<Array<AuditReportTemplate>> {
    var response = await this.getReportTemplates(0, 100);
    var templates = response.items as Array<AuditReportTemplate>;
    while (response.links.next) {
      response = await this.getReportTemplates(response.page + 1, 100);
      templates = templates.concat(response.items);
    }
    return templates;
  }

  getAuditConfigs(page: number = 0, pageSize: number = 1000): Promise<PaginatedList<AuditReportConfig>> {
    return http.get<PaginatedList<AuditReportConfig>>('/rails/audit/configs', { params: { "page": page, "page-size": pageSize }})
      .then(response => response.data);
  }

  async fetchAllConfigs(): Promise<Array<AuditReportConfig>> {
    var response = await this.getAuditConfigs(0, 100);
    var templates = response.items as Array<AuditReportConfig>;
    while (response.links.next) {
      response = await this.getAuditConfigs(response.page + 1, 100);
      templates = templates.concat(response.items);
    }
    return templates;
  }

  createAuditConfig(config: AuditReportConfig): Promise<AuditReportConfig> {
    return http.post<AuditReportConfig>('/rails/audit/configs', config)
      .then(response => response.data);
  }

  getAuditConfig(configId: string): Promise<AuditReportConfig> {
    return http.get<AuditReportConfig>(`/rails/audit/configs/${configId}`)
      .then(response => response.data);
  }

  updateAuditConfig(config: AuditReportConfig): Promise<AuditReportConfig> {
    return http.put<AuditReportConfig>(`/rails/audit/configs/${config.id}`, config)
      .then(response => response.data);
  }

  deleteAuditConfig(configId: string): Promise<any> {
    return http.delete<void>(`/rails/audit/configs/${configId}`);
  }

  getAuditIssueSummaries(): Promise<Array<AuditIssueSummary>> {
    return http.get<Array<AuditIssueSummary>>('/rails/audit/summaries')
      .then(response => response.data);
  }

  getAuditIssues(configId: string, acknowledged?: boolean, page: number = 0, pageSize: number = 1000): Promise<PaginatedList<AuditIssue>> {
    var params: any = { "page": page, "page-size": pageSize };
    if ((acknowledged !== undefined) && (acknowledged !== null)) {
      params = { "acknowledged": acknowledged, ...params };
    }

    return http.get<PaginatedList<AuditIssue>>(`/rails/audit/configs/${configId}/issues`, { params: params })
      .then(response => response.data);
  }

  getAuditIssue(issueId: string): Promise<AuditIssue> {
    return http.get<AuditIssue>(`/rails/audit/issues/${issueId}`)
      .then(response => response.data);
  }

  updateAuditIssue(issueId: string, acknowledged: boolean): Promise<AuditIssue> {
    const request = { acknowledged: acknowledged };
    return http.put<AuditIssue>(`/rails/audit/issues/${issueId}`, request)
      .then(response => response.data);
  }

  deleteAuditIssue(issueId: string): Promise<any> {
    return http.delete<void>(`/rails/audit/issues/${issueId}`);
  }
}

const instance = new AuditReportService();
export default instance;

File: rail-module/rail-service/src/main/java/com/hillayes/rail/service/AuditReportService.java
package com.hillayes.rail.service;

import com.hillayes.commons.Strings;
import com.hillayes.commons.jpa.Page;
import com.hillayes.exception.common.MissingParameterException;
import com.hillayes.exception.common.NotFoundException;
import com.hillayes.rail.audit.AuditReportTemplate;
import com.hillayes.rail.domain.AuditIssue;
import com.hillayes.rail.domain.AuditIssueSummary;
import com.hillayes.rail.domain.AuditReportConfig;
import com.hillayes.rail.errors.AuditReportConfigAlreadyExistsException;
import com.hillayes.rail.repository.AccountTransactionRepository;
import com.hillayes.rail.repository.AuditIssueRepository;
import com.hillayes.rail.repository.AuditReportConfigRepository;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.inject.Any;
import jakarta.enterprise.inject.Instance;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import lombok.extern.slf4j.Slf4j;

import java.time.Instant;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;

@ApplicationScoped
@Transactional
@Slf4j
public class AuditReportService {
    @Inject
    @Any
    Instance<AuditReportTemplate> reportTemplates;

    @Inject
    AuditReportConfigRepository auditReportConfigRepository;

    @Inject
    AuditIssueRepository auditIssueRepository;

    @Inject
    AccountTransactionRepository accountTransactionRepository;

    public Page<AuditReportTemplate> getAuditTemplates(int page, int pageSize) {
        log.info("Listing audit report templates [page: {}, pageSize: {}]", page, pageSize);
        List<AuditReportTemplate> templates = reportTemplates.stream()
            .sorted(Comparator.comparing(AuditReportTemplate::getName, String::compareToIgnoreCase))
            .toList();
        return Page.of(templates, page, pageSize);
    }

    public Page<AuditReportConfig> getAuditConfigs(UUID userId, int page, int pageSize) {
        log.info("Listing audit report configs [userId: {}, page: {}, pageSize: {}]", userId, page, pageSize);
        return auditReportConfigRepository.findByUserId(userId, page, pageSize);
    }

    public AuditReportConfig createAuditConfig(UUID userId, AuditReportConfig config) {
        log.info("Creating audit report config [userId: {}, name: {}]", userId, config.getName());
        String newName = Strings.trimOrNull(config.getName());
        if (newName == null) {
            throw new MissingParameterException("AuditReportConfig.name");
        }

        auditReportConfigRepository.findByUserAndName(userId, newName)
            .ifPresent(existing -> { throw new AuditReportConfigAlreadyExistsException(existing); });

        config.setUserId(userId);
        config.setName(newName);
        return auditReportConfigRepository.save(config);
    }

    public AuditReportConfig getAuditConfig(UUID userId, UUID configId) {
        log.info("Getting audit report config [userId: {}, configId: {}]", userId, configId);
        return auditReportConfigRepository.findByIdOptional(configId)
            .filter(config -> config.getUserId().equals(userId))
            .orElseThrow(() -> new NotFoundException("AuditReportConfig", configId));
    }

    public AuditReportConfig updateAuditConfig(UUID userId, UUID configId, AuditReportConfig update) {
        log.info("Updating audit report config [userId: {}, configId: {}, name: {}]", userId, configId, update.getName());

        AuditReportConfig original = getAuditConfig(userId, configId);

        String newName = Strings.trimOrNull(update.getName());
        if (newName == null) {
            throw new MissingParameterException("AuditReportConfig.name");
        }

        auditReportConfigRepository.findByUserAndName(userId, newName)
            .filter(existing -> !existing.getId().equals(configId))
            .ifPresent(existing -> { throw new AuditReportConfigAlreadyExistsException(existing); });

        original.setVersion(update.getVersion());
        original.setDisabled(update.isDisabled());
        original.setName(update.getName());
        original.setDescription(update.getDescription());
        original.setTemplateName(update.getTemplateName());
        original.setReportSource(update.getReportSource());
        original.setReportSourceId(update.getReportSourceId());
        original.setUncategorisedIncluded(update.isUncategorisedIncluded());

        // update existing or add new parameters
        update.getParameters().values().forEach(param ->
            original.getParameter(param.getName()).ifPresentOrElse(
                p -> p.setValue(param.getValue()),
                () -> original.addParameter(param.getName(), param.getValue())
            )
        );

        // delete missing parameters
        original.getParameters().keySet().stream()
            .filter(paramName -> !update.getParameters().containsKey(paramName))
            .toList()
            .forEach(original::removeParameter);

        return auditReportConfigRepository.save(original);
    }

    public void deleteAuditConfig(UUID userId, UUID configId) {
        log.info("Deleting audit report config [userId: {}, configId: {}]", userId, configId);
        auditReportConfigRepository.delete(getAuditConfig(userId, configId));
    }

    public void deleteAllAuditConfigs(UUID userId) {
        log.info("Deleting audit report configs [userId: {}]", userId);
        auditReportConfigRepository.deleteByUserId(userId);
    }

    public List<AuditIssueSummary> getIssueSummaries(UUID userId) {
        log.info("Get audit issue summaries [userId: {}]", userId);
        return auditIssueRepository.getIssueSummaries(userId);
    }

    public Page<AuditIssue> getAuditIssues(UUID userId, UUID configId, Boolean acknowledged,
                                           int page, int pageSize) {
        log.info("Listing audit report issues [userId: {}, configId: {}, acknowledged: {}, page: {}, pageSize: {}]",
            userId, configId, acknowledged, page, pageSize);
        getAuditConfig(userId, configId);
        return auditIssueRepository.findByConfigId(configId, acknowledged, page, pageSize);
    }

    public AuditIssue getAuditIssue(UUID userId, UUID issueId) {
        log.info("Get audit report issue [userId: {}, issueId: {}]", userId, issueId);
        return auditIssueRepository.findByIdOptional(issueId)
            .filter(issue -> issue.getUserId().equals(userId))
            .orElseThrow(() -> new NotFoundException("AuditIssue", issueId));
    }

    public AuditIssue updateIssue(UUID userId, UUID issueId, boolean acknowledged) {
        log.info("Updating audit issue [issueId: {}, acknowledged: {}]", issueId, acknowledged);
        AuditIssue issue = auditIssueRepository.findByIdOptional(issueId)
            .filter(i -> i.getUserId().equals(userId))
            .orElseThrow(() -> new NotFoundException("AuditIssue", issueId));
        issue.setAcknowledgedDateTime(acknowledged ? Instant.now() : null);
        return auditIssueRepository.save(issue);
    }

    public void deleteIssue(UUID userId, UUID issueId) {
        log.info("Deleting audit issue [issueId: {}]", issueId);
        auditIssueRepository.delete(auditIssueRepository.findByIdOptional(issueId)
            .filter(i -> i.getUserId().equals(userId))
            .orElseThrow(() -> new NotFoundException("AuditIssue", issueId)));
    }
}
